src/irc/BridgedClient.ts

```diff
@@ -302,51 +308,78 @@ export class BridgedClient extends EventEmitter {
         return this.inst.disconnect(reason, textReason);
     }
 
+    /**
+     * Determines if a nick name already exists.
+     */
+    public async checkNickExists(nick: string): Promise<boolean> {
+        try {
+            // We don't care about the return value of .whois().
+            // It will return null if the user isn't defined.
+            return (await this.whois(nick)) !== null;
+        }
+        catch (error) {
+            if (error.message === "Cannot find nick on whois.") {
+                return false;
+            }
+            throw error;
+        }
+    }
+
     /**
      * Change this user's nick.
      * @param {string} newNick The new nick for the user.
      * @param {boolean} throwOnInvalid True to throw an error on invalid nicks
      * instead of coercing them.
      * @return {Promise<String>} Which resolves to a message to be sent to the user.
      */
-    public changeNick(newNick: string, throwOnInvalid: boolean): Promise<string> {
-        let validNick = newNick;
-        try {
-            validNick = this.getValidNick(newNick, throwOnInvalid);
-            if (validNick === this.nick) {
-                return Promise.resolve(`Your nick is already '${validNick}'.`);
-            }
+    public async changeNick(newNick: string, throwOnInvalid: boolean): Promise<string> {
+        this.log.info(`Trying to change nick from ${this.nick} to ${newNick}`);
+        const validNick = this.getValidNick(newNick, throwOnInvalid);
+        if (validNick === this.nick) {
+            throw Error(`Your nick is already '${validNick}'.`);
         }
-        catch (err) {
-            return Promise.reject(err);
+        if (validNick !== newNick) {
+            // Don't "suggest" a nick.
+            throw Error("Nickname is not valid");
+        }
+
+        if (await this.checkNickExists(validNick)) {
+            throw Error(
+                `The nickname ${newNick} is taken on ${this.server.domain}.` +
+                "Please pick a different nick."
+            );
         }
+
+        return await this.sendNickCommand(validNick);
+    }
+
+    private sendNickCommand(nick: string): Promise<string> {
         if (!this.unsafeClient) {
-            return Promise.reject(new Error("You are not connected to the network."));
+            throw new Error("You are not connected to the network.");
```

Comment:
In all other cases this function is returning a Promise and not throwing. If this was from m, I wanted to make this an async function, which is tricky with the callbacks below.
```suggestion
            return Promise.reject(new Error("You are not connected to the network."));
```

Reply:
Well, if we should probably make the func be async if it returns a promise anyway. Let me do that.